File: modules/webrtc-core/src/main/java/com/shterneregen/securelan/webrtc/runtime/RtcEngine.java
package com.shterneregen.securelan.webrtc.runtime;

import com.shterneregen.securelan.common.model.rtc.RtcSessionMode;
import com.shterneregen.securelan.common.model.rtc.RtcSignalEnvelope;
import com.shterneregen.securelan.webrtc.event.RtcEvent;

import java.util.function.Consumer;

public interface RtcEngine extends AutoCloseable {
    RtcRuntimeStatus status();

    void startSession(
            String sessionId,
            String localPeer,
            String remotePeer,
            RtcSessionMode mode,
            String dataChannelLabel,
            Consumer<RtcSignalEnvelope> outboundSignalConsumer,
            Consumer<RtcEvent> eventConsumer
    );

    void handleRemoteSignal(RtcSignalEnvelope signal, Consumer<RtcSignalEnvelope> outboundSignalConsumer, Consumer<RtcEvent> eventConsumer);

    void sendData(String sessionId, String payload, Consumer<RtcEvent> eventConsumer);

    void closeSession(String sessionId, Consumer<RtcEvent> eventConsumer);

    @Override
    default void close() {
    }
}

File: modules/webrtc-core/src/main/kotlin/com/shterneregen/securelan/webrtc/runtime/RtcEngine.kt
package com.shterneregen.securelan.webrtc.runtime

import com.shterneregen.securelan.common.model.rtc.RtcSessionMode
import com.shterneregen.securelan.common.model.rtc.RtcSignalEnvelope
import com.shterneregen.securelan.webrtc.event.RtcEvent
import java.util.function.Consumer

interface RtcEngine : AutoCloseable {
    fun status(): RtcRuntimeStatus

    fun startSession(
        sessionId: String?,
        localPeer: String?,
        remotePeer: String?,
        mode: RtcSessionMode?,
        dataChannelLabel: String?,
        audioCaptureDeviceId: String?,
        videoCaptureDeviceId: String?,
        outboundSignalConsumer: Consumer<RtcSignalEnvelope>,
        eventConsumer: Consumer<RtcEvent>,
    )

    fun handleRemoteSignal(
        signal: RtcSignalEnvelope,
        outboundSignalConsumer: Consumer<RtcSignalEnvelope>,
        eventConsumer: Consumer<RtcEvent>,
    )

    fun sendData(sessionId: String?, payload: String?, eventConsumer: Consumer<RtcEvent>)

    fun closeSession(sessionId: String?, eventConsumer: Consumer<RtcEvent>)

    override fun close() {
    }
}
